### PR TITLE
Move coreos-iso-embed-ignition to coreos-installer subcommands

### PR DESCRIPTION
**Pull request by bgilbert**

### CODE REVIEW

File: src/embed.rs [hunk 1]
```diff
@@ -0,0 +1,268 @@
...
+}
+
+struct EmbedArea<'a> {
+    file: &'a File,
+    offset: u64,
+    length: usize,
+}
+
+impl<'a> EmbedArea<'a> {
+    fn for_file(file: &'a mut File) -> Result<Self> {
```
Comment (lucab) [inline src/embed.rs]: Can we document somewhere the metadata we are expecting?
From my reading of the code, we read the last 24B of this 32KiB area at the beginning of the disk and we expect the following:
 * `[u8; 8]` - magic "coreiso+"
 * `u64` little-endian - offset (from ISO byte 0?) of the cpio archive
 * `u64` little-endian - length of the cpio archive

File: src/embed.rs [hunk 2]
```diff
@@ -0,0 +1,268 @@
+// Copyright 2019 CoreOS, Inc.
```
[the remark this answers is not in the document]
Comment (bgilbert) [reply src/embed.rs]: It handles all of the `iso` subcommands.  `iso.rs`?

File: src/embed.rs [hunk 3]
```diff
@@ -0,0 +1,268 @@
...
+}
+
+/// Make a gzipped CPIO archive containing the specified Ignition config.
+fn make_cpio(ignition: &[u8]) -> Result<Vec<u8>> {
+    let mut result = Cursor::new(Vec::new());
+    let encoder = GzBuilder::new()
+        .mtime(0)
+        .write(&mut result, Compression::best());
+    let mut input_files = vec![(
+        NewcBuilder::new(FILENAME).mode(0o100_644),
```
[the remark this answers is not in the document]
Comment (bgilbert) [reply src/embed.rs]: It's `S_IFREG` | 0644.

File: src/cmdline.rs [hunk 4]
```diff
@@ -127,10 +146,83 @@ pub fn parse_args() -> Result<Config> {
...
+        )
         .get_matches();
 
     match app_matches.subcommand() {
         ("install", Some(matches)) => parse_install(&matches),
+        ("iso", Some(iso_matches)) => match iso_matches.subcommand() {
+            ("embed", Some(matches)) => parse_iso_embed(&matches),
+            ("show", Some(matches)) => parse_iso_show(&matches),
+            ("remove", Some(matches)) => parse_iso_remove(&matches),
+            _ => bail!("unrecognized subcommand"),
```
[the remark this answers is not in the document]
Comment (bgilbert) [reply src/cmdline.rs]: This shouldn't ever be reached, because the command-line parser should error out before we get here.

**Comment by bgilbert** [comment]
Updated.

File: src/embed.rs (hunk 3 above)
[the remark this answers is not in the document]
Comment (lucab) [reply src/embed.rs]: Ah, I missed that. Thanks for adding the inline comment.

**Comment by lucab** [comment]
Code looks fine to me. @jlebon do you want to have a pass too? Otherwise I'll take care of stamping in a few days.

@bgilbert I just have an high-level doubt: we are nesting a JSON inside a CPIO inside a GZ, and we lookup for a single well-known filename. Why did you put a cpio in the middle there instead of directly compressing the JSON? Is it for future-proofing or is there more behind it?

**Comment by bgilbert** [comment]
We're generating an additional initramfs image for the kernel to unpack into the initramfs at runtime.  The embed area is a block of padding concatenated to the end of the initrd file.

**Pull request merged**